File: typescript/packages/core/src/sandbox/exec.ts
import { assertNever } from "../assert-never";
import type { ArtifactRef } from "../log";
import type { ToolRun } from "../loop/types";
import { err, ok, type Result } from "../result";
import type { ArtifactStore } from "../store/artifacts";
import type {
  ExecOptions,
  ExecOutput,
  Failure,
  SandboxContext,
  SandboxSession,
  Stale,
} from "./protocol";

// The sandbox layer over an adapter's exec (spec/api.json ExecResult): both
// streams go to one artifact as they arrive, in arrival order, while only head and tail
// previews stay in memory. The artifact is kept only when a stream outgrew its preview.

export type ExecResult = {
  readonly exit_code: number;
  /** Head and tail preview. */
  readonly stdout: string;
  readonly stderr: string;
  readonly truncated: boolean;
  /** Present exactly when truncated: the complete stdout and stderr. */
  readonly full_output?: ArtifactRef;
};

/** Bytes of head, and of tail, each stream keeps for its preview. */
export const PREVIEW_BYTES = 4096;

const text = new TextDecoder();

/** The first and last `keep` bytes of a stream, and how many there were. */
class Preview {
  #head: Uint8Array = new Uint8Array(0);
  #tail: Uint8Array = new Uint8Array(0);
  total = 0;

  constructor(readonly keep: number) {}

  add(chunk: Uint8Array): void {
    this.total += chunk.length;
    const room = this.keep - this.#head.length;
    if (room > 0) this.#head = concat(this.#head, chunk.subarray(0, room));
    const rest = chunk.subarray(Math.max(room, 0));
    if (rest.length > 0)
      this.#tail = concat(this.#tail, rest).slice(-this.keep);
  }

  get truncated(): boolean {
    return this.total > this.#head.length + this.#tail.length;
  }

  toString(): string {
    const omitted = this.total - this.#head.length - this.#tail.length;
    const marker = omitted > 0 ? `\n[... ${omitted} bytes omitted ...]\n` : "";
    return `${text.decode(this.#head)}${marker}${text.decode(this.#tail)}`;
  }
}

function concat(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a);
  out.set(b, a.length);
  return out;
}

async function pump(
  stream: AsyncIterable<Uint8Array>,
  preview: Preview,
  write: (chunk: Uint8Array) => void,
): Promise<void> {
  for await (const chunk of stream) {
    preview.add(chunk);
    write(chunk);
  }
}

export type ExecFailure =
  | Failure<"timeout" | "invalid_path" | "unavailable">
  | Stale;

/**
 * An exec's outcome as a tool run: a timeout or a
 * transport failure after dispatch is uncertain (effect_unknown), never a result; a refused
 * fence sent nothing.
 */
export function toolRunOf(result: Result<ExecResult, ExecFailure>): ToolRun {
  if (result.ok)
    return {
      kind: "done",
      output: JSON.stringify(result.value),
      isError: result.value.exit_code !== 0,
    };
  const { code, message } = result.error;
  switch (code) {
    case "timeout":
      return { kind: "unknown", reason: "timeout" };
    case "unavailable":
      return { kind: "unknown", reason: "transport_error" };
    case "stale_epoch":
    case "cleanup_claim_lost":
      return { kind: "not_sent" };
    case "invalid_path":
      return { kind: "done", output: message, isError: true };
    default:
      return assertNever(code);
  }
}

/** Runs `command` in the session and spills its output at the source. */
export async function execute(
  session: SandboxSession,
  command: readonly string[],
  context: SandboxContext,
  options: ExecOptions,
  artifacts: ArtifactStore,
  keep: number = PREVIEW_BYTES,
): Promise<Result<ExecResult, ExecFailure>> {
  const started = await session.exec(command, context, options);
  if (!started.ok) return started;
  const { timeoutMs } = options;
  const collected = collect(started.value, artifacts, keep);
  if (timeoutMs === undefined) return collected;
  // After a timeout nobody awaits it; a stream that breaks later is not a crash.
  collected.catch(() => undefined);
  // Once the deadline passes the outcome is a timeout, whatever the process does next: the
  // kill is best effort and unconfirmed, so the effect stays unknown.
  const deadline = Promise.withResolvers<Result<ExecResult, ExecFailure>>();
  const timer = setTimeout(() => {
    deadline.resolve(
      err({ code: "timeout", message: `no exit within ${timeoutMs} ms` }),
    );
    void session.terminate(options.processKey, context);
  }, timeoutMs);
  try {
    return await Promise.race([collected, deadline.promise]);
  } finally {
    clearTimeout(timer);
  }
}

async function collect(
  output: ExecOutput,
  artifacts: ArtifactStore,
  keep: number,
): Promise<Result<ExecResult, ExecFailure>> {
  const sink = artifacts.sink();
  const out = new Preview(keep);
  const errs = new Preview(keep);
  try {
    await Promise.all([
      pump(output.stdout, out, sink.write),
      pump(output.stderr, errs, sink.write),
    ]);
  } catch (error) {
    sink.abort();
    throw error;
  }
  const exitCode = await output.exit_code;
  const truncated = out.truncated || errs.truncated;
  const base = {
    exit_code: exitCode,
    stdout: out.toString(),
    stderr: errs.toString(),
    truncated,
  };
  if (!truncated) {
    sink.abort();
    return ok(base);
  }
  const full = sink.finish();
  return ok({
    ...base,
    full_output: { ...full, media_type: "application/octet-stream" },
  });
}

File: python/src/threads/sandbox/exec.py
"""The sandbox layer's exec: an adapter's `ExecOutput` streams
become an `ExecResult` of head and tail previews. Every chunk is spilled to the artifact store
as it arrives, so full output is never held in host memory; it becomes `full_output` exactly
when a stream outgrows its preview, and is durable before the result is returned."""

import asyncio
from collections.abc import AsyncIterator, Mapping, Sequence
from dataclasses import dataclass, field

from threads.log import ArtifactRef, Spill
from threads.result import Err, Ok
from threads.sandbox.protocol import (
    NO_ENV,
    ExecOutput,
    ExecResult,
    SandboxContext,
    SandboxError,
    SandboxSession,
)
from threads.store.spill import Spill as Sink

_MARKER = "\n[output truncated: {n} bytes in all; the full output is in full_output]\n"


@dataclass(frozen=True, slots=True)
class Command:
    """One exec: the command and the options spec/api.json names."""

    argv: Sequence[str]
    process_key: str
    """The call's effect key: the process group's durable identity."""
    cwd: str = "/workspace"
    env: Mapping[str, str] = field(default_factory=lambda: NO_ENV)
    timeout_ms: int | None = None
    stdin: bytes | None = None


@dataclass(slots=True)
class _Preview:
    """One stream's preview: all of it while it fits, else its head and tail."""

    limits: Spill
    whole: bytearray = field(default_factory=bytearray)
    tail: bytearray = field(default_factory=bytearray)
    head: bytes = b""
    total: int = 0

    @property
    def truncated(self) -> bool:
        return self.total > self.limits.threshold_bytes

    def add(self, chunk: bytes) -> None:
        fitted = not self.truncated
        self.total += len(chunk)
        if fitted:
            self.whole += chunk
            if not self.truncated:
                return
            # This chunk outgrew the preview: keep only the head and the tail from now on.
            self.head, self.tail = bytes(self.whole[: self.limits.head_bytes]), self.whole
            self.whole = bytearray()
        else:
            self.tail += chunk
        del self.tail[: max(0, len(self.tail) - self.limits.tail_bytes)]

    def text(self) -> str:
        if not self.truncated:
            return self.whole.decode("utf-8", "replace")
        head, tail = self.head.decode("utf-8", "ignore"), self.tail.decode("utf-8", "ignore")
        return head + _MARKER.format(n=self.total) + tail


_stopping: set[asyncio.Task[object]] = set()
"""Best-effort terminates still running after their exec answered timeout."""


async def run_exec(
    session: SandboxSession, command: Command, context: SandboxContext, spill: Sink, limits: Spill
) -> Ok[ExecResult] | Err[SandboxError]:
    """Runs `command` and consumes both streams at once into `spill`, in arrival order (as a
    terminal shows them). The deadline covers the start too. At the deadline the answer is
    timeout, whatever the process does next: terminate is asked for in the background and
    never awaited, so a stop that can't finish can't hold the exec, and a process the stop
    killed never reads as a normal exit."""
    out, err = _Preview(limits), _Preview(limits)
    deadline = None if command.timeout_ms is None else command.timeout_ms / 1000
    try:
        async with asyncio.timeout(deadline):
            started = await _start(session, command, context)
            if isinstance(started, Err):
                await spill.discard()
                return started
            await asyncio.gather(
                _drain(started.value.stdout, out, spill), _drain(started.value.stderr, err, spill)
            )
            code = await started.value.exit_code
    except TimeoutError:
        await spill.discard()
        stop = asyncio.create_task(_terminate(session, command.process_key, context))
        _stopping.add(stop)
        stop.add_done_callback(_stopping.discard)
        return Err(SandboxError("timeout", f"exec timed out after {command.timeout_ms} ms"))
    truncated = out.truncated or err.truncated
    full = None
    if truncated:
        sha = await spill.commit()
        full = ArtifactRef(sha256=sha, bytes=spill.written, media_type="application/octet-stream")
    else:
        await spill.discard()
    return Ok(ExecResult(code, out.text(), err.text(), truncated, full))


async def _start(
    session: SandboxSession, command: Command, context: SandboxContext
) -> Ok[ExecOutput] | Err[SandboxError]:
    return await session.exec(
        command.argv,
        context,
        process_key=command.process_key,
        cwd=command.cwd,
        env=command.env,
        timeout_ms=command.timeout_ms,
        stdin=command.stdin,
    )


async def _terminate(session: SandboxSession, process_key: str, context: SandboxContext) -> object:
    # Its answer settles nothing here: the effect is already effect_unknown and parks.
    return await session.terminate(process_key, context)


async def _drain(stream: AsyncIterator[bytes], preview: _Preview, spill: Sink) -> None:
    async for chunk in stream:
        preview.add(chunk)
        await spill.write(chunk)
